addons/account_asset/account_asset.py

```diff
@@ -1,351 +1,325 @@
-                    elif i == undone_dotation_number:
-                        amount = (residual_amount * asset.method_progress_factor) / total_days * (total_days - days)
+                    if sequence == 1:
+                        amount = (residual_amount * self.method_progress_factor) / total_days * days
+                    elif sequence == undone_dotation_number:
+                        amount = (residual_amount * self.method_progress_factor) / total_days * (total_days - days)
         return amount
 
-    def _compute_board_undone_dotation_nb(self, cr, uid, asset, depreciation_date, total_days, context=None):
-        undone_dotation_number = asset.method_number
-        if asset.method_time == 'end':
-            end_date = datetime.strptime(asset.method_end, '%Y-%m-%d')
+    @api.multi
+    def _compute_board_undone_dotation_nb(self, depreciation_date, total_days):
+        undone_dotation_number = self.method_number
+        if self.method_time == 'end':
+            end_date = datetime.strptime(self.method_end, '%Y-%m-%d')
             undone_dotation_number = 0
             while depreciation_date <= end_date:
-                depreciation_date = (datetime(depreciation_date.year, depreciation_date.month, depreciation_date.day) + relativedelta(months=+asset.method_period))
+                depreciation_date = (datetime(depreciation_date.year, depreciation_date.month, depreciation_date.day) + relativedelta(months=+self.method_period))
                 undone_dotation_number += 1
-        if asset.prorata:
+        if self.prorata and self.category_id.type == 'purchase':
             undone_dotation_number += 1
         return undone_dotation_number
 
-    def compute_depreciation_board(self, cr, uid, ids, context=None):
-        depreciation_lin_obj = self.pool.get('account.asset.depreciation.line')
-        currency_obj = self.pool.get('res.currency')
-        for asset in self.browse(cr, uid, ids, context=context):
-            if asset.value_residual == 0.0:
-                continue
-            posted_depreciation_line_ids = depreciation_lin_obj.search(cr, uid, [('asset_id', '=', asset.id), ('move_check', '=', True)],order='depreciation_date desc')
-            old_depreciation_line_ids = depreciation_lin_obj.search(cr, uid, [('asset_id', '=', asset.id), ('move_id', '=', False)])
-            if old_depreciation_line_ids:
-                depreciation_lin_obj.unlink(cr, uid, old_depreciation_line_ids, context=context)
-
-            amount_to_depr = residual_amount = asset.value_residual
-            if asset.prorata:
-                depreciation_date = datetime.strptime(self._get_last_depreciation_date(cr, uid, [asset.id], context)[asset.id], '%Y-%m-%d')
+    @api.multi
+    def compute_depreciation_board(self):
+
+        posted_depreciation_line_ids = self.env['account.asset.depreciation.line'].search([('asset_id', '=', self.id), ('move_check', '=', True)], order='depreciation_date desc')
```

Comment:
Why ('asset_id', '=', self.id) ???
The method is expected to contain a multiple record.